Improve symlog axis ticks

schtandard opened the pull request:
<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary
<!-- Please provide at least 1-2 sentences describing the pull request in detail
(Why is this change required?  What problem does it solve?) and link to relevant
issues and PRs.
Also please summarize the changes in the title, for example "Raise ValueError on
non-numeric input to set_xlim" and avoid non-descriptive titles such as "Addresses
issue #8576".
-->

In the current implementation, `symlog` axes only get major ticks by default. Depending on the data range, this can lead to too few or even no ticks at all. When "subticks" between decades are specified manually, those are sometimes labeled in an unintended manner (when they are at a decade).

This PR re-implements the tick placement and formatting for `symlog` axes to remedy these problems. If I am not mistaken, this closes #17402, closes #18757 and closes #25994.

I am submitting this as a PR now to get some general feedback on the approach and pointers if I need to consider anything regarding style or conventions that I missed. If I get a thumbs up, I would polish some things off (like making `_SymmetricalLogUtil`'s methods work with `np.array`s and have `.firstdec()` cache results) before merging.

## My approach for the tick locator

I was aiming for `SymmetricalLogLocator` to behave identically to `LogLocator` as long as the data range is outside of the linear part of the scale and extend this reasonably to the linear part if it's present. What I came up with is this (I'm describing the positive half axis here for simplicity, everything works the same on the negative side):

- Major ticks at decades, where the first decade to be considered for labelling must be at least half the length of a decade from 0 (otherwise there would be infinitely many labeled decades close to 0).
- That decade receives the _decade number_ 1, the next larger one 2 and so on. 0 gets the "decade" number 0. Values between those decades get interpolated decade numbers.
- We now have an obvious definition of the number of decades in the data range and can basically copy the logic from `LogLocator`, the only caveat being that 0 should always receive a tick if it is in the range.
- In the linear part of the scale, we add the same minor ticks below the first decade as below any other and add the next lower decade as a minor tick. That is, if with base b one gets minor ticks 2 to b-1 between larger decades one gets 1 to b-1 below the first one, effectively producing linear ticks between the first decade and 0.

`LogFormatter` (which is also used for `symlog` axes) is adapted accordingly.

- I added a separate field `_firstsublabels` for placing sublabels below the first decade.
- As with `log` axes it can sometimes happen that there is only one major tick with unlabeled minor ticks around it. This is fine (one can still identify the unlabeled positions) as long as the major tick is at some decade and not at 0, which can happen with `symlog` axes. In that case, we need to label at least one of the minor ticks to allow reading values off the axis.

Since both the locator and the formatter need the calculation of the first decade I placed the code regarding this into a helper class `_SymmetricalLogUtil`.

Finally, `SymmetricalLogScale` is adapted to actually use the new functionality by default.

I checked all parameter ranges I could think of and this seems to work out quite nicely.
<details>
<summary>This is how it looks.</summary>

![showcase_new](https://github.com/matplotlib/matplotlib/assets/23075534/2a16868f-d5a5-4722-9b93-6874ced84216)

</details>
<details>
<summary>For comparison, this is how it used to look.</summary>

![showcase_current](https://github.com/matplotlib/matplotlib/assets/23075534/efa57ea8-8046-438b-b4c8-20506a618458)

</details>
<details>
<summary>Code to reproduce.</summary>

```python
import numpy as np
from matplotlib import pyplot as plt

def showcase_scales(axs, *plotargs, **plotkwargs):
    for ax, scale in zip(axs, ['linear', 'symlog', 'log']):
        ax.set_yscale(scale)
        ax.yaxis.grid()
        ax.plot(*plotargs, **plotkwargs)

values = [
    np.linspace(3, 150),
    np.linspace(20, 60),
    np.linspace(0.5, 12),
    np.linspace(0.5, 5),
    np.linspace(0.6, 3),
    np.linspace(1e-4, 1e-2),
    np.linspace(0, 40),
    np.linspace(-0.8, 0.9),
    np.linspace(-4, 4),
    np.linspace(-22, 22),
    np.linspace(0, 0),
    np.linspace(1, 1),
    np.linspace(2, 2),
    np.linspace(10, 10),
    np.linspace(13, 13),
]
fig, axs = plt.subplots(len(values), 3, squeeze=False, sharex=True,
                        figsize=(9, 2 * len(values)))
for a, v in zip(axs, values):
    showcase_scales(a, v)
axs[0, 0].set_title('linear')
axs[0, 1].set_title('symlog')
axs[0, 2].set_title('log')
fig.tight_layout()
fig.savefig('showcase.png')
```

</details>

## Requested feedback

There are some details regarding which I would like feedback/guidance from you.

- When computing `stride` (i.e. the number of decades between major ticks) a possible forced tick at 0 is not considered. This may lead to an off-by-one error regarding `numticks` (i.e. a `stride` value smaller by one would have been possible) but I feel like preventing this is not worth the complication of taking it into account. Do you agree?
- `subs=None` is equivalent to `subs='auto'` in `LogLocator` but to `subs=(1.0,)` in `SymmetricalLogLocator`. Is that on purpose? I left it like that.
- For some reason, `base` came _after_ `linthresh` in `SymmetricalLogLocator`'s signature, contrary to most other occurences. In adding `linscale` I reordered those. While this is strictly a backwards incompatible change (whereas I just appended `linscale` everywhere else), I don't expect it to have ever been used in an incompatible way: providing `base` only make sense when `transform` (the first argument) is not given, so it's most probably always been don by keyword argument, though it is in principle possible to say `SymmetricalLogLocator(None, None, 2, 10)` to get `SymmetricalLogLocator(linthresh=2, base=10)`. Should I switch it back for backwards compatibility or do we care more about consistent argument order?
- `LogFormatter` threw away the sign of any value, even though it is used for negative values with `symlog` axes. This was no problem because `LogFormatterMathtext` overwrites `__call__` and preserves the sign, so it didn't actually show up on any axis (with the default configuration). I fixed this. Is it fine to just leave it in this PR or should it be a separate one?
- In `test_ticker.py` the test `TestSymmetricalLogLocator.test_extending()` relies on `rcParams['axes.autolimit_mode']` being set to `round_numbers` but this is not set up in the test itself. I assume that this configuration is left in place by some previous test, though I can't tell where. Is this by design or should an `rc_context` be used, like in some other tests?
- The checklist below tells me to note API changes (which the new `linscale` argument to the locator and the formatter as well as the changed default value of `SymmetricalLogScale`'s `subs` argument qualify as, I think) with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes), but the linked section does not seem to exist. What do I need to do?

## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

schtandard commented:
Ah, about the tests: Not all tests ran succesfully, but all the ones that fail also do so on `main`, so I'm confident this is not due to my changes.

schtandard commented:
Leaving this comment to ping you guys, as it's been more than a week since I opened this PR. :-)

schtandard commented:
Leaving another ping-comment, as it's been almost six weeks.

QuLogic commented:
Sorry there hasn't been much discussion here. I've put this on [the agenda for tomorrow's meeting](https://hackmd.io/l9vkn_T4RSmk147H_ZPPBA).

QuLogic commented:
Sorry I meant to comment last week after the meeting, but it seemed to have been left here without going through.

In general, we are in favour of this change. However, we would like to see that it be behind a flag of some sort, either a) an argument to the class init, or b) an rcParam, or c) both. Sometime in the future, it may make sense to flip that flag (which I would be in favour of), but it should be tested out a bit first.

> * When computing `stride` (i.e. the number of decades between major ticks) a possible forced tick at 0 is not considered. This may lead to an off-by-one error regarding `numticks` (i.e. a `stride` value smaller by one would have been possible) but I feel like preventing this is not worth the complication of taking it into account. Do you agree?

I think I've seen an issue related to this somewhere and will try to find it.

> * `subs=None` is equivalent to `subs='auto'` in `LogLocator` but to `subs=(1.0,)` in `SymmetricalLogLocator`. Is that on purpose? I left it like that.

It was probably an artifact of whoever implemented things, but then was refactored to share code and now it seems inconsistent. But now it can't easily change for backwards compatibility reasons.

> * For some reason, `base` came _after_ `linthresh` in `SymmetricalLogLocator`'s signature, contrary to most other occurences. In adding `linscale` I reordered those. While this is strictly a backwards incompatible change (whereas I just appended `linscale` everywhere else), I don't expect it to have ever been used in an incompatible way: providing `base` only make sense when `transform` (the first argument) is not given, so it's most probably always been don by keyword argument, though it is in principle possible to say `SymmetricalLogLocator(None, None, 2, 10)` to get `SymmetricalLogLocator(linthresh=2, base=10)`. Should I switch it back for backwards compatibility or do we care more about consistent argument order?

Unfortunately we do, so it will have to be reverted. In some cases, we might want to start making arguments keyword-only, but that does require going through a deprecation cycle.

> * `LogFormatter` threw away the sign of any value, even though it is used for negative values with `symlog` axes. This was no problem because `LogFormatterMathtext` overwrites `__call__` and preserves the sign, so it didn't actually show up on any axis (with the default configuration). I fixed this. Is it fine to just leave it in this PR or should it be a separate one?

It's probably fine here, but if you have some easy way to show that difference outside of the other changes, it might be quicker for review as a separate PR.

> * In `test_ticker.py` the test `TestSymmetricalLogLocator.test_extending()` relies on `rcParams['axes.autolimit_mode']` being set to `round_numbers` but this is not set up in the test itself. I assume that this configuration is left in place by some previous test, though I can't tell where. Is this by design or should an `rc_context` be used, like in some other tests?

This is set by the classic style, which is the default for all tests, though new tests should start using `mpl20` and in that case you would need to set the `rcParam` yourself. There's not really a need for `rc_context` unless you are mixing more than one call in a test, as all rcParams are reset before each test.

> * The checklist below tells me to note API changes (which the new `linscale` argument to the locator and the formatter as well as the changed default value of `SymmetricalLogScale`'s `subs` argument qualify as, I think) with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes), but the linked section does not seem to exist. What do I need to do?

I think this might have moved [here](https://matplotlib.org/devdocs/devel/api_changes.html). @story645 I guess we need to update the pull request template for some of the recent refactors?

schtandard commented:
Thanks for your comments! I think I have everything to make final adjustments now. I'm quite busy at the moment, but I will try and get that done soon.

rcomer commented:
Does this also close #10369?

schtandard commented:
Hi @melissawm, thanks for the ping and sorry for the long delay; a stressfull period of my life dragged on longer than I had expected. It finally concluded last month, though, so I hope I'll now be able to wrap this up. In fact, I already took a quick look recently to see what's left to do. I'll have another crack and try to incorporate all the things mentioned above, hopefully without as much delay this time.

schtandard committed:
Do not discard sign of tick value

schtandard committed:
Improve some log ticker details

- Use the same logic to calculate numticks in LogFormatter.set_locs() as
  later in LogLocator.tick_values().
- Do not hard-code n_request == 9 in minor tick check.
- Correct a type in an explanatary comment (< instead of <=).
- Actually calculate the largest possible stride, as outlined in the
  comments.
- The check for lonely ticks at the end of LogLocator.tick_values()
  assumed that major ticks are not also minor ticks. Fix that logic.

schtandard committed:
Improve symlog ticker

The new SymmetricalLogLocator emulates the behavior of LogLocator in the
logarithmic region and extends its behavior to the linear part in a
(hopefully) reasonable manner.

The change is hidden behind a new rcParam (and corresponding class
parameters) for now.

schtandard committed:
Update symlog tests

schtandard committed:
Rename new rcParam

schtandard committed:
Streamline subtick determination

- For SymLogLocator, do not use a hard-coded limit preventing minor
  ticks. Instead, always use the same behavior.
- For LogLocator, revert to the old behavior with the hard-coded limit
  of 10 available major ticks.

schtandard committed:
Fix missing stub parameter

timhoffm commented:
Sure, we can discuss. I’ll put my opinion in already here.

So far we have done several changes to ticks without backs compatibility options, e.g. #29698, #29054, #18754, #13069 (and possibly more).

We haven’t seen any complaints on these improvements/fixes of tick positions so likely nobody relies on them (or they can cope with changes). Therefore, flags just for the sake of keeping old behavior in tick locations seem like unnecessary complexity.

schtandard commented:
Has any conclusion been reached here?

timhoffm commented:
There hasn't been any further discussion / feedback. @QuLogic do you remember whether there have been strong opinions? If there's no feedback within a week, I'd wield my authority as API lead and decide that we don't need a backward-compatibility parameter, based on precedence for tick position changes mentioned in https://github.com/matplotlib/matplotlib/pull/27310#issuecomment-4185103072.

schtandard committed:
Remove legacy behavior

schtandard committed:
Add new parameters backward-compatibly

schtandard committed:
Note API changes

schtandard commented:
Alright, let's see of we can get this over the finish line now. I removed the legacy behavior and added two files noting the API change. Here is a summary of the signature changes:

- `SymmetricLogScale()` now has default value `subs='auto'` rather than `subs=None`.
- `LogFormatter()` has new parameter `linscale`.
- `SymmetricalLogLocator()` has new parameters `numticks` and `linscale`.
- `SymmetricalLogLocator.set_params()` has new parameters `base`, `linthresh`, `linscale`.

Two notes:

- The first change is the only backward-incompatible one, as all others only add new parameters at the end of the function signature. I don't think there is a better option, though. The old default was to exclude minor ticks, which is not good. As noted [above](https://github.com/matplotlib/matplotlib/pull/27310#issuecomment-4068214414), this might also be a remnant from previous "bad API".
- In `SymmetricalLogLocator()`, I placed the new `numticks` parameter as a keyword-only argument at the end of the signature, analogously to `LogLocator()`. This is now inconsistent with `SymmetricalLogLocator.set_params()`, where `numticks` was already there as a normal argument and I needed to place the new arguments after it for backward-compatibility. There is no consistent, backward-compatible way available.

I tried to follow the [API guidelines](https://matplotlib.org/devdocs/devel/api_changes.html#announce-new-and-deprecated-api) for announcing the changes, however I wasn't entirely sure how much detail I should go into. Let me know if this is lacking.

I also read that I should add `.. versionadded:: 3.12` directives in the docstrings for the new parameters, however I am unsure how to do this in the case of the `linscale` parameters, as they are documented together with `base` and `linthresh`. Is there a way to mark only `linscale` as new without separating the documentation from the other parameters? Should I also add a `.. versionchanged:: 3.12` directive to the `subs` parameter with the new default value? Is there a way to note _what_ changed?

timhoffm commented on doc/api/next_api_changes/behavior/27310-BW.rst:
```diff
@@ -0,0 +1,6 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The tick placement for ``symlog`` axes has been improved. It now also supports
+minor ticks and `SymmetricalLogScale` uses ``subs='auto'`` (minor ticks if
```
```suggestion
minor ticks and `.SymmetricalLogScale` uses ``subs='auto'`` (minor ticks if
```

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -939,9 +1033,9 @@ class LogFormatter(Formatter):
-        parameter must be supplied here.
+    linthresh, linscale : None or float, default: None
+        If a symmetric log scale is in use, its ``linthresh`` and ``linscale``
+        parameters must be supplied here.
```
```suggestion
        parameters must be supplied here.
        
        .. versionadded:: 3.12
            The *linscale* parameter.
```

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -1001,6 +1098,22 @@ def set_label_minor(self, labelOnlyBase):
         self.labelOnlyBase = labelOnlyBase
 
+    @property
+    def _symlog(self):
```
```suggestion
    def _is_symlog(self):
```

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -1001,6 +1098,22 @@ def set_label_minor(self, labelOnlyBase):
+                                                   linscale=self._linscale)
+            return True
+        try:
+            self._symlogutil = _SymmetricalLogUtil(self.axis.get_transform())
```
Do we need the deferred initialitation? This seems to make the logic more complex comared to initializing in __init__

schtandard replied:
You are right, I don't think we need this. I think I might have been worried about `set_base()`, but this was not actually addressed by the deferred instantiation. Will revise.

schtandard replied:
Actually, I take it back: The fallback using `self.axis.get_transform()` cannot work at initialisation, as `self.axis` is not yet set there. This only showed up in the `axes3d` tests, so I didn't catch it. Will add a `symlog` test that actually shows minor ticks for the future. (Or maybe I'll just change `test_symlog()`. We are changing it anyways..)

timhoffm replied:
Ok, in that case, we should make the deferred initiatlization more explicit. Doing it as as side-effect of _is_symlog may be surprising.

Either override `set_axis` and do the initialization there. Or make `_symlogutil` a functools.cached_property.

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,100 @@ def _set_format(self):
+            self.linscale = linscale
+        else:
+            raise ValueError("Either transform, or all of base, linthresh and "
+                             "linscale must be provided.")
```
Simpler:
```suggestion
    def __init__(self, base, linthresh, linscale):
        self.base = base
        self.linthresh = linthresh
        self.linscale = linscale
    
    @classmethod
    def from_transform(transform):
        return cls(
            base=transform.base,
            linthresh=transform.linthresh,
            linscale=transform.linscale,
        )
```

schtandard replied:
Hmm, I see the point, but then we need to move the logic checking `transform` to `SymmetricalLogLocator()` in [line 2812](https://github.com/matplotlib/matplotlib/pull/27310/changes#diff-ffc88dbb928475229d0fc50ae389ef3f0dabc9c5b8abf42b6eb7a0af28afdb66R2812), no? I think that is the signature I modeled this one after.

timhoffm replied:
Correct. Still I think this is the better approach. `_SymmetricalLogUtil` collects logic that is used in `SymmetricalLogLocator` and `LogFormatter`. Thus we shouldn't push down the awkward parameter concurrency of `SymmetricalLogLocator`.

`_SymmetricalLogUtil.__init__` should just have the parameter it actually needs.  It doesn't care about transforms. Determining whether the values come from individual parameters or from the transform is not of its concern.  You may or may not add the `from_transform` classmethod as a convenience. Thinking about it maybe even not.

The implementation in `SymmetricalLogLocator` is quite straight forward:

```python
        if transform is not None:
            if any(p is not None for p in (base, linthresh, linscale):
                 rase ValueError(...)
            base = transform.base,
            linthesh = transform.linthesh,
           linscale = transform.linscale,
        self._symlogutil = _SymmetricalLogUtil(base, linthresh, linscale)
```

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -1001,6 +1098,22 @@ def set_label_minor(self, labelOnlyBase):
+                                                   linscale=self._linscale)
+            return True
+        try:
+            self._symlogutil = _SymmetricalLogUtil(self.axis.get_transform())
```
```suggestion
            self._symlogutil = _SymmetricalLogUtil.from_transform(self.axis.get_transform())
```

schtandard committed:
Fix rst syntax

Co-authored-by: Tim Hoffmann <2836374+timhoffm@users.noreply.github.com>

schtandard committed:
Rename _symlog to _is_symlog

schtandard committed:
Instantiate _symlogutil in init()

schtandard committed:
Relay changed base value to _symlogutil

schtandard committed:
Add versioning directives

schtandard committed:
Create _symlogutil before using set_base()

schtandard committed:
Delay _symlogutil instantiation

We rely on the axis transformation when no explicit symlog parameters
are given, but that is not known at initialisation.

schtandard committed:
Use minor ticks in symlog test

schtandard committed:
Update test output

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,100 @@ def _set_format(self):
+        if val > 1:
+            val = self.linscale + self._log_b(val)
+        else:
+            val *= self.linscale
```
Let's write it like this: 
```suggestion
        sign, val = np.sign(val), np.abs(val)
        if val > self.linthresh:  # log regime
            val = self.linscale + self._log_b(val / self.linthresh)
        else:  # linear regime
            val = val / self.linthresh * self.linscale
```

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,100 @@ def _set_format(self):
 
+class _SymmetricalLogUtil:
+    """
+    Helper class for working with symmetrical log scales.
```
I still struggle to understand what exactly this class does. It's possibly a combination of vague documentation and naming. I comment on specific aspects below.

schtandard replied:
This is a fair point. I added an explanation to the docstring that hopefully clears up much of the confusion.

timhoffm replied:
👍 The docstring is a substantial improvement. We'll go through the details one-by-one.

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,100 @@ def _set_format(self):
+
+    def pos(self, val):
+        """
+        Calculate the normalized position of the value on the axis.
```
"normalized" seems imprecise. It's rather "scaled", but a special internal helper scale.

Possibly also the function `scaled()`?

schtandard replied:
To me, "normalized" seems totally appropriate, maybe that's my physics-backround. `scaledpos()` or something like that would be possible, of course.

I refrained from renaming the methods for now, as I felt it's best to re-evaluate with the extended docstring. If you still prefer longer names, I'm open for that of course. Maybe `normpos()`, `inv_normpos()`, `decnum()`, `inv_decnum()` and `first_decnum()`? Though I do like the visual parallel of `log` and `dec` in `LogFormatter.set_locs()` and `LogLocator.tick_values()`/`SymmetricalLogLocator.tick_values()`...

timhoffm replied:
The docstring is a substantial improvement. 👍  I asked AI to check logical consistency and correct use of terminology (detailed answer below).

Overall the terminology topic sticks.

- "decade" is a range not a single point. So probably you are right with "*powers* of base" as defining the decade boundaries.

- "normalized": Overall we are doing a mapping from the data space to the "symlog space", where decades have a with of 1 - or in other words: the distance between neighboring powers is 1. Would it be reasonable to define this with the term "symlog space", in analogy to "log space"?

---

<details>
<summary>Detailed answer</summary>
Overall, the text is logically coherent and the mathematical construction is internally consistent. The main weaknesses are not in the mathematics but in the precision of the terminology and a few places where the wording obscures the underlying concepts.

Overall assessment

The exposition introduces three related quantities:

logarithm (used for ordinary log axes),
normalized position (a geometric coordinate on the transformed axis),
decade number (an abstract numbering used for tick selection).

These are kept distinct until the final paragraph, where their relationships are explained. That structure is good.

The main issues are:

"normalized position" is somewhat misleading,
"decade number" is overloaded with the ordinary meaning of "decade",
several statements could be made more mathematically precise.
1. "Normalized position"

This is the weakest terminology.

You define it as

its distance from 0 normalized to the size of one decade in the logarithmic regime

but this is not normalization in the sense most readers expect.

Normally "normalized" suggests

mapping to a standard range,
scaling to unit norm,
or making dimensionless.

Here only the third interpretation applies: you divide by a characteristic length.

So the quantity is really

position measured in units of one logarithmic decade.

It is dimensionless, but it is not normalized to a fixed interval.

The terminology is therefore not incorrect, but it is potentially misleading.

Better names might be

scaled position
decade-scaled position
logarithmic-scale coordinate
normalized distance (slightly better)

If you keep "normalized position", I'd explicitly state

"Here 'normalized' means expressed in units of one logarithmic decade; it is not confined to a fixed range."

2. "Decade number"

This is mathematically consistent but slightly confusing.

Normally

decade

means

the interval between consecutive integer logarithms.

For example,

10
2
 to 10
3

is a decade.

Here you instead define

decade number

as a continuous coordinate extending the logarithm.

That is reasonable, but then you say

We assign it the decade number 1 and number larger powers of base consecutively.

This is effectively defining a new coordinate

d(x)

rather than numbering decades.

It is really an

extended logarithmic coordinate,
logarithmic index,
symlog logarithmic coordinate,

rather than a "decade number."

Not wrong, just slightly overloaded terminology.

3. Infinite powers near zero

This paragraph is logically sound.

The logarithmic approach cannot work in the linear part, as there are infinitely many powers of base near 0...

This is true.

The only stylistic issue is that

infinitely many powers of base near 0

initially sounds odd.

More explicitly,

there are infinitely many positive powers b
n
 approaching 0 as n→−∞

is clearer.

4. Acceptable tick positions

The logic here is good.

You define

normalized position,
threshold 0.5,
acceptable powers.

The only slight awkwardness is

We choose a power of base to be an acceptable tick position if its distance from 0 on the axis is at least half the size of one decade...

Later this becomes

normalized position greater than 0.5.

Those are equivalent, but the first definition sounds geometric while the second is algebraic.

I'd define the normalized position first, then state the threshold immediately afterwards.

5. Coordinate vs axis position

You alternate between

coordinate

and

position on the axis.

These are close but not identical.

For example,

normalized position of a coordinate

is slightly awkward.

Either consistently use

coordinate,

or

axis position.
6. "Distance from 0"

This deserves a little clarification.

You write

distance from 0 on the axis

Later,

coordinates in the linear regime have a normalized position between -linscale and linscale

Since distances are nonnegative, the signed normalized position is actually not a distance.

So the definition is slightly inconsistent.

You appear to mean

signed distance

or

signed axis coordinate measured relative to 0.

Otherwise readers may wonder why distances become negative.

7. Final paragraph

This is the strongest part.

The relationships are clearly separated:

normalized position = logarithm + constant
decade number = logarithm + constant
special parameter choice ⇒ all three coincide

That progression is mathematically clean.

I would only tighten

identical to the logarithm of the coordinate

to

identical to log
b
	​

(∣x∣) up to a constant offset

since the logarithm is only defined for positive values and you've introduced a symmetric construction.

Minor wording issues

Negative decade numbers receive negative decade numbers

This is clearly a typo.

It should be

Negative coordinates receive negative decade numbers...

or

Coordinates on the negative side are assigned negative decade numbers...

reasonably extend it

Somewhat subjective.

Perhaps

naturally extend it

or

extend it in a consistent manner.

first decade

This momentarily sounds like the interval between 1 and b, whereas you actually mean

the smallest positive major tick.

Summary

The mathematics is internally consistent and the construction is well motivated. The main issues are terminological rather than logical:

"Normalized position" is technically defensible but may mislead readers into expecting a bounded quantity. "Scaled position" or an explicit explanation that it is measured in units of one logarithmic decade would reduce ambiguity.
"Decade number" is a reasonable abstraction but overloads the established meaning of "decade." It functions more like an extended logarithmic coordinate than a label for decades.
The only outright logical inconsistency is the use of "distance" to define a quantity that later takes negative values. Referring instead to a signed distance or signed position would resolve this cleanly.
</details>

schtandard replied:
Alright, I thought about it a bunch and tried to come up with better names, though it's difficult to find one that avoids potential erroneous associations altogether. I saw similar pitfalls with the AI suggestions. In the end I settled on _axis position_ instead of _normalized position_ (keeping _decade number_).

I also renamed the methods to more explicit names and slightly revised the dostring once more.

timhoffm replied:
I'm not completely happy with that naming because we have the *axis coordinate system* which is normalized to [0, 1]. "axis position" sounds too close to that;

Possibly a visualization is helpful as well. AFAICS we have this:
```
Coordinate:
  |---+-------+-------+--linear--+--linear--+-------+-------+---|
    10^-5   10^-4 >   10^-3        0        10^3    10^4    10^5

Decade number:
  |---+-------+-------+--linear--+--linear--+-------+-------+---|
     -3      -2      -1          0          1       2       3

"Axis position" (ls=linscale):
  |---+-------+-------+--linear--+--linear--+-------+-------+---|
   -(ls+2) -(ls+1)   ls          0         ls      ls+1    ls+2

Axis coordinate
  |---+-------+-------+--linear--+--linear--+-------+-------+---|
  0                             0.5                             1
```
So what you call "axis position" is a shifted and scaled axis coordinate. Maybe that can inform a better name?

Possibly also define the purpose first and then the technical definition. Some thing like

> We define two helper coordinate systems:
> - *decade number* serves to easily identify the powers: Powers are located at integer decade numbers, and the smallest needed power is located at decade number.
> - *axis position* is a bridge between *decade number* and *axis coordinates*: It solves the problem to define the visual size of the linear region. By defining the size of the linear region relative to a decade. It's connected to the axis coordinate by a shift and scale so that equal distances in "axis position" result in equal visual distances.

schtandard replied:
Your visualization is correct, except that in general _linthresh_ is not a power of _base_ (in fact, it is 2 by default while _base_ is 10). A more complete try to visualize this might look like this:
```plain
Coordinate:
                      |---------linear----------|
  |-----+-------+-------+----------+----------+-------+-------+-----|
      10^-5   10^-4   10^-3        0        10^3    10^4    10^5

Decade number:
                      |---------linear----------|
  |-----+-------+-------+----------+----------+-------+-------+-----|
       -3      -2      -1          0          1       2       3

"Axis position" (ls=linscale):
                      |---------linear----------|
  |---+-------+-------+------------+------------+-------+-------+---|
     (ls+2) -(ls+1)   ls           0            ls      ls+1    ls+2

Axis coordinate
                      |---------linear----------|
  |--------------------------------+--------------------------------|
  0                               0.5                               1
```
Of course, the linear regime may also be smaller than the [-1, 1] decade number interval. I also want to explicitly point out that the relation between _decade number_ and _coordinate_ is _not_ linear inside the full range marked `linear` here. That is only true for _axis position_ and _axis coordinate_. (For _decade number_, that range is the [-1, 1] interval.)

I hadn't thought about the "axis coordinate", but the association is actually quite fitting, since, as you noted, the _axis position_ is connected to it through shifting and scaling, while this is not true for the other two coordinate systems. I added an initial paragraph to the docstring as you suggested. (I find it quite hard to determine how much explanation is sufficient and how much is too much to the point of being less digestible. I'm relying on your guidance here.) Do you think it's clear like this or do we need to look for a different name than _axis position_?

timhoffm replied:
This is good now. :+1: Please add the ASCII drawing to the docstring. I think a visualization really helps a lot. Yes the docs are a bit verbose here, but the concepts are complicated an I think it really helps navigating what's going on.

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,100 @@ def _set_format(self):
+            val *= self.linscale
+        return sign * val
+
+    def unpos(self, val):
```
```suggestion
    def unscale(self, val):
```
?

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,100 @@ def _set_format(self):
+            val /= self.linscale
+        return sign * val * self.linthresh
+
+    def firstdec(self):
```
```suggestion
    def first_decade(self):
```

Actually writing out is not enough. Returning an exponent and the value and calling it "decade" is confusing.
Also naming the result of power "power is incorrect. `power()` calculates base to the power of exp (or power). If some value should be called power, it would be the exponent.

I stumbled over this in `self._symlogutil.firstdec()[1]` which is really not readable. Either find a good name that indicates the two values or make two separate functions.

schtandard replied:
You are right that returning two values is not ideal, it now returns just the second value. The first one now has its own method `firstdec_exp()`, though it is only used in `firstdec()` so maybe we should just move it into one function..?

schtandard replied:
Regarding the word "power": I'm not a native speaker, but [I believe](https://en.wiktionary.org/wiki/power#English:_exp) that "power" denotes the result of the calculation, i.e. "128 and 1024 are powers of 2". Nevertheless, I have renamed to avoid this issue. (Though I do use the word a bit in the extended class docstring.)

timhoffm replied:
I think you are right about power naming.

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,100 @@ def _set_format(self):
+        firstpow = np.power(self.base, firstexp)
+        return firstexp, firstpow
+
+    def dec(self, val):
```
```suggestion
    def decade_number(self, val):
```
?

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,100 @@ def _set_format(self):
+        """
+        Calculate the decade number of the value. The first decade to have a
+        position (given by pos) of at least 0.5 is given the number 1, the
+        value 0 is given the decade number 0.
```
Make it a little clearer that *decade number*  is a concept we define.

```suggestion
        Calculate the *decade number* of the value.
        
        We incrementally number the decades, starting with 1 for the
        first decade to have a scaled position (given by pos) of at least 0.5.
        The value 0 is given the decade number 0.
```

schtandard replied:
I'd say this is addressed by the extended class docstring. Or do you think we should re-emphasize it in the method docstring?

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -1032,43 +1143,48 @@ def set_locs(self, locs=None):
-                        - math.floor(math.nextafter(lmin, -math.inf)))
-            numdec = abs(lmax - lmin)
+            numdec = lmax - lmin
+            numticks = math.floor(lmax) - math.ceil(lmin) + 1
```
The structure is parallel but the naming minrdec vs.lmin is quite different. Can we make this more similar?

schtandard replied:
Done. I renamed to `mindec`/`maxdec` and `minlog`/`maxlog`. Did the same in the `Locator`s, where the rounded integer values `imindec`/`ìmaxdec` or `iminlog`/`imaxlog` also appear.

timhoffm reviewed (commented):
Sorry, on a closer look, I find this really hard to follow. I wasn't able to follow the logic and validate the numerics. I think we first have to work on structure, naming and documentation.

schtandard committed:
Simplify _SymmetricalLogUtil.init()

schtandard committed:
Make _symlogutil instantiation more explicit

schtandard committed:
Simplify calculation syntax

schtandard committed:
Remove docstring remnants

schtandard committed:
Explain normalized position and decade number

schtandard committed:
Streamline firstdec usage

This also makes the implementation of dec() and undec() completely
analogous to that of pos() and unpos().

schtandard committed:
Rename variables consistently

schtandard commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,119 @@ def _set_format(self):
+
+    def firstdec_exp(self):
+        """Calculate the exponent of the first decade."""
+        return np.ceil(self._log_b(self.unpos(0.5)))
```
I want to point out that the value of 0.5 here is somewhat arbitrary. It controls the minimum distance of the first major tick from the 0 tick. Here it is half the size of one decade in the logarithmic regime. Depending on the axis, this might still be a bit close, I guess, so maybe a value closer to 1 could also be reasonable.

timhoffm replied:
Doesn't this also depend on the number of decades? If I have 1-2 decases a smaller value may be better compared to 8 decases.
Overall, just try some cases and see what looks good.

schtandard replied:
It doesn't strongly depend on the number of decades (except when there are so few that the fallback tick placement becomes active, but then everything is different): Major ticks in the logarithmic regime never get denser than one per decade, so the ticks in the linear regime should roughly match that spacing.

Thinking about it again, I did come up with a reason for the value of 0.5 (which I probably already used back in the day): With this threshold, the linear minor tick distance (for a _base_ of 10) is always at least as large as the smallest minor tick distance in the logarithmic regime. So it actually is not too arbitrary. I added this motivation to the docstring.

schtandard committed:
Update stub patterns

schtandard committed:
Remove spurious stub default values

schtandard commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,119 @@ def _set_format(self):
+                np.log2(x) if self.base == 2 else
+                np.log(x) / np.log(self.base))
+
+    def pos(self, val):
```
Not sure if I mentioned it before: This method is never actually used. I mostly included it for consistency and because I thought it helped understanding `unpos()`. Only `unpos()` is used, and that only once (in `firstdec_exp()`).

timhoffm replied:
Yes, I've noticed it's unused. It's ok to add it for consistency.

Could possilby be `data_to_symlog` or `data_to_symlog_scale`. 
Likewise the inverse `symlog_to_data`.

schtandard committed:
Choose clearer method names

Also revise the docstring a bit.

schtandard committed:
Motivate threshold for acceptable tick positions

schtandard committed:
Describe function of helper coordinates first

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -908,6 +908,136 @@ def _set_format(self):
+    def __init__(self, base, linthresh, linscale):
+        self.base = base
+        self.linthresh = linthresh
+        self.linscale = linscale
```
Data ownership is not trivial here. These values can come in three ways:
- in LogFormatter as explicit parameters
- in LogFormatter from the axis transform
- in SymmetricalLogLocator as explicit paramters

I'm considering dropping owned state from `_SymmetricalLogUtil` so that we don't have to cope with updates or deferred initialization. After all, this is just a bunch of (parametrized) functions.

Ideally, we could make them a mixin, relying on the parent object state, but I don't overlook the case of the transform. An alternative would be to replace `base, linthres, linscale` by a `data_source` object.

So  that you could do `_SymmertricalLogUtil(self)` where `self` would be the Formatter or Locator.

Would something like this makes sense?

schtandard replied:
I think the mixin makes most sense and I gave it a shot. Let me know if this is the structure you were imagining.

Now that I dropped the case where LogFormatter inherits the axis properties, we don't have the complication of deferred access to the values, but even with that the mixin would work, I think, as we only ever use the methods after checking `self._is_symlog`, which checks for their existence.

timhoffm replied:
Yes, now we're getting somewhere 🎉

schtandard committed:
Include ASCII drawing relating coordinate systems

schtandard committed:
Instantiate symlog formatter with all parameters

schtandard commented:
I wanted to again understand why the deferred instantiation is necessary in the first place and got quite confused: Surely we should just instantiate the formatter with _linthresh_ and _linscale_ in `SymmetricalLogScale.set_default_locators_and_formatters()` (cf. [#5615500](https://github.com/matplotlib/matplotlib/pull/27310/commits/56155003fd5fe7a6c19c4743175ca4c386cd5e3a)), no? It's been a while, so I can't remember if there was a reason for not doing that (or not relying on the formatter being set there), but I can't think of one now.

schtandard committed:
Convert _SymmetricalLogUtil to a mixin

schtandard committed:
Make mixin methods private

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -979,7 +1118,14 @@ def __init__(self, base=10.0, labelOnlyBase=False,
+
+    @property
+    def _is_symlog(self):
+        return all(p is not None for p in (self._base, self._linthresh, self._linscale))
```
`self._base` cannot be None for a LogFormatter.
```suggestion
        return all(p is not None for p in (self._linthresh, self._linscale))
```

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -908,7 +908,144 @@ def _set_format(self):
-class LogFormatter(Formatter):
+class _SymmetricalLogMixin:
+    """
+    Common methods for working with symmetrical log scales.
```
```suggestion
    Common methods for working with symmetrical log scales.
    
    The mixin is used in `.LogFormatter` and `.SymmetricalLogLocator`. It relies
    on the private attibutes ``_base``, ``_linthres``, and ``_linscale``.
```

timhoffm commented on lib/matplotlib/scale.py:
```diff
@@ -621,10 +624,16 @@ def __init__(self, axis=None, *, base=10, linthresh=2, subs=None, linscale=1):
         axis.set_major_locator(SymmetricalLogLocator(self.get_transform()))
-        axis.set_major_formatter(LogFormatterSciNotation(self.base))
+        axis.set_major_formatter(LogFormatterSciNotation(base=self.base,
+                                                         linthresh=self.linthresh,
```
There's an asymmetry here between the locator and the formatter: The locator is initialized from the transform, the formatter is initialized from the individual parameters base, linthresh, linscale, which in turn reach in to the transform through the above defined properties.

I suggest we converge to one pattern here: The underlying data are the individual parameters, so let's use them. This also means we can get rid of the transform parameter and thus simplify the API (possibly do this as a follow-up since we need to go through a deprecation anyway).

schtandard replied:
> This also means we can get rid of the transform parameter and thus simplify the API (possibly do this as a follow-up since we need to go through a deprecation anyway).

By "a follow-up" you mean a separate PR?

timhoffm replied:
Yes

timhoffm commented:
> I wanted to again understand why the deferred instantiation is necessary in the first place and got quite confused: Surely we should just instantiate the formatter with _linthresh_ and _linscale_ in `SymmetricalLogScale.set_default_locators_and_formatters()` (cf. [#5615500](https://github.com/matplotlib/matplotlib/pull/27310/commits/56155003fd5fe7a6c19c4743175ca4c386cd5e3a)), no? It's been a while, so I can't remember if there was a reason for not doing that (or not relying on the formatter being set there), but I can't think of one now.

That sounds reasonable.

schtandard committed:
Document where mixin is used

schtandard committed:
Avoid spurious check

schtandard committed:
Use individual parameters rather than transform

schtandard commented on lib/matplotlib/ticker.pyi:
```diff
@@ -100,14 +100,22 @@ class ScalarFormatter(Formatter):
 
-class LogFormatter(Formatter):
+class _SymmetricalLogMixin:
+    def _val2axpos(self, val: float) -> float: ...
+    def _axpos2val(self, val: float) -> float: ...
+    def _firsttickval(self) -> float: ...
+    def _val2decnum(self, val: float) -> float: ...
+    def _decnum2val(self, val: float) -> float: ...
```
Before, these methods were public (while the class was not), which is why I listed them here. Not sure now if the class should be removed altogether..? Or should I leave it and add `._log_b()` for completeness? (This has always been private, so was not included here.)

timhoffm replied:
Yes, please keep them. I suspect otherwise type checkers may have a harder time interpreting the functions in which they are used.

timhoffm commented on lib/matplotlib/ticker.py:
```diff
@@ -979,7 +1121,15 @@ def __init__(self, base=10.0, labelOnlyBase=False,
+    @property
+    def _is_symlog(self):
+        # self._base cannot be None, so we don't need to check it.
+        return all(p is not None for p in (self._linthresh, self._linscale))
```
Simpler:
```suggestion
        return self._linthresh is not None and self._linscale is not None
```
The comment on base is not needed as, for the context of LogFormatter it's obvious that base is a general property.

timhoffm commented:
```
/home/circleci/project/lib/matplotlib/ticker.py:docstring of matplotlib.ticker.LogFormatter:1: WARNING: py:class reference target not found: matplotlib.ticker._SymmetricalLogMixin [ref.class]
/home/circleci/project/lib/matplotlib/ticker.py:docstring of matplotlib.ticker.SymmetricalLogLocator:1: WARNING: py:class reference target not found: matplotlib.ticker._SymmetricalLogMixin [ref.class]
```
This warning is an issue with sphinx inheritance diagrams and private (and thus not publically documented/linkable) base classes. Please fix by adding the warnings to https://github.com/matplotlib/matplotlib/blob/main/doc/missing-references.json

timhoffm reviewed (commented):
This is almost ready. Only minor fixes need.

Thanks for bearing the intensive review process!

schtandard committed:
Simplify _is_symlog test

schtandard committed:
Fix changed attribute name

schtandard committed:
Register missing _SymmetricalLogMixin warning

schtandard commented:
> Thanks for bearing the intensive review process!

:+1: I think it was worth it, thanks.

Any verdict on the `_SymmetricalLogMixin` stub patterns? (Not sure why I can't link to my comment above.)

I don't really understand the failing Azure checks. For the Codecov one I could add more tests, if you want. The PR cleanliness we will handle by squashing once everything is finished, I assume?

timhoffm commented:
> Any verdict on the `_SymmetricalLogMixin` stub patterns? (Not sure why I can't link to my comment above.)

What exactly, I don't find the context. If you can't link, please copy-paste.
 
> I don't really understand the failing Azure checks. For the Codecov one I could add more tests, if you want. The PR cleanliness we will handle by squashing once everything is finished, I assume?

The windows failure is a flaky timeout - unrelated to the PR. You could add some tests for the `_SymmetricalLogMixin` functions. They would be a good addition to documenting what they do. Other than that, I wouldn't go hunting for coverage.

schtandard commented:
> > Any verdict on the `_SymmetricalLogMixin` stub patterns? (Not sure why I can't link to my comment above.)
> 
> What exactly, I don't find the context. If you can't link, please copy-paste.

Ah, [apparently](https://github.com/orgs/community/discussions/10369) I somehow started a new review and need to submit it before it becomes visible to you.. Can you see it now?

Just in case: I'm referring to [lines 103--108 of `ticker.pyi`](https://github.com/matplotlib/matplotlib/pull/27310/changes#diff-9e5a9e5464f7ae8c955576c148959962faf4fe31043e293f161c067be8c0f788R103-R108) and commented this:

> Before, these methods were public (while the class was not), which is why I listed them here. Not sure now if the class should be removed altogether..? Or should I leave it and add `._log_b()` for completeness? (This has always been private, so was not included here.)

schtandard committed:
Add tests for the mixin methods

QuLogic commented on doc/release/next_whats_new/improved_symlog_ticks.rst:
```diff
@@ -0,0 +1,7 @@
+Improved tick placement for ``symlog`` axes
+-------------------------------------------
```
Can you add a sample plot for the what's new note?

schtandard replied:
Yes, I added one. Have a look if you think it does a good job.

QuLogic replied:
Since the text says ticks will be placed identically to `log` axes, can you rearrange it a bit to show that? Currently, the limits are so different that that isn't so clear.

schtandard replied:
I'm not sure if I understand you correctly. Do you mean adding `log` plots alongside the `symlog` plots showing the logarithmic part of the respective axis range (i.e. everything above 2)? Or do you mean add a `symlog` axis containing only the logarithmic range and put a `log` axis with the same range next to it? Or something else?

QuLogic replied:
Currently, the plot looks like this:
<img width="550" height="450" alt="image" src="https://github.com/user-attachments/assets/911e3154-4b17-4698-803c-1a37558b9782" />
On the right plot, the log scale is from ~5e-1 to ~7e0, while on the left we have (for positive) 0 to ~3e2.

What I suggest is adjusting the limits such that the right plot has the top at ~3e2 as well, and the bottom such that the 10⁰, 10¹ and 10² are aligned, so that it is (hopefully) clear that the log scale is the same up until the point it transitions to linear.

schtandard replied:
I'm still not quite sure I understand. Both of these are symlog plots, so adjusting the limits in the way you describe would just result in two identical plots, no? My intention with the second plot was to show the labeling of minor ticks, which is why it has a smaller range. I assumed (maybe wrongly), that log plots would be sufficiently familiar to anyone looking at symlog plots that the logarithmic part of the symlog axis would be readily recognized.

If we want to make the log part behavior more explicit, we could add a log plot next to it, maybe like this.

<img width="640" height="480" alt="log_comparison_display" src="https://github.com/user-attachments/assets/e8b1f216-6220-49cf-830c-e7fee32526c7" />

Would that provide the clarification you are looking for?

QuLogic commented on doc/release/next_whats_new/improved_symlog_ticks.rst:
```diff
@@ -0,0 +1,7 @@
+The placement of ticks for ``symlog`` axes has been improved. Ticks are now
+placed identically to ``log`` axes in the logarithmic part with a reasonable
+extension of this behavior to the linear part of the axis. Axes with too few
+ticks or spurious ticks should now be avoided.
```
This last sentence is a bit ambiguous as it could be read as either "will now be avoided by this new implementation" or "should now be avoided by users of this new implementation".

QuLogic commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -1377,19 +1377,22 @@ def test_fill_between_interpolate_nan():
+    axs[1].set_yscale('symlog', linthresh=1, linscale=2)
 
 
 @image_comparison(['symlog2.pdf'], remove_text=True, style='_classic_test')
```
As it appears you've updated this test image, you can change the style to `mpl20`.

QuLogic commented on lib/mpl_toolkits/mplot3d/tests/baseline_images/test_axes3d/scale3d_all_scales.png:
Please run `oxipng` on these updated PNG images.

QuLogic commented on lib/matplotlib/scale.py:
```diff
@@ -599,6 +599,9 @@ class SymmetricalLogScale(ScaleBase):
         8 logarithmically spaced minor ticks between each major tick.
 
+        .. versionchanged:: 3.12
+            The default value is now ``'auto'``.
```
`'auto'` is not documented or mentioned here.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -908,7 +908,147 @@ def _set_format(self):
+        Equal distances in axis position correspond to equal visual distances. It is
+        connected to the axis coordinate by a shift and scale such that the value 0 has
+        the axis position 0 and consecutive logarithmic decades differ by 1 in axis
+        position.
```
This is misaligned:
```suggestion
    We define two helper coordinate systems:

    - *decade number* serves to easily identify the powers of *base* that may serve as
      major tick positions. Such powers are located at integer decade numbers, and the
      smallest needed power is located at decade number 1.
    - *axis position* helps to gauge the visual distance between values on the axis.
      Equal distances in axis position correspond to equal visual distances. It is
      connected to the axis coordinate by a shift and scale such that the value 0 has
      the axis position 0 and consecutive logarithmic decades differ by 1 in axis
      position.
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -908,7 +908,147 @@ def _set_format(self):
+
+    axis coordinate: |--------------------------------+--------------------------------|
+                     0                               0.5                               1
+                     |--------log--------|---------linear----------|--------log--------|
```
This won't render too well and should probably go in a code block.
```suggestion
    ::

                       |--------log--------|--------linear---------|--------log--------|
           coordinate: |-----+-------+-------+---------+---------+-------+-------+-----|
                           10^-5   10^-4   10^-3       0       10^3    10^4    10^5

        decade number: |-----+-------+-------+---------+---------+-------+-------+-----|
                            -3      -2      -1         0         1       2       3

        axis position: |---+-------+-------+-----------+-----------+-------+-------+---|
                          (ls+2) -(ls+1)   ls          0           ls      ls+1    ls+2

      axis coordinate: |-------------------------------+-------------------------------|
                       0                              0.5                              1
                       |--------log--------|--------linear---------|--------log--------|
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -908,7 +908,147 @@ def _set_format(self):
+    def _firsttickval(self):
+        """
+        Calculate the value of the first acceptable (positive) tick position. We define
+        this to the first power of *base* with an axis position of at least 0.5. This
```
```suggestion
        this to be the first power of *base* with an axis position of at least 0.5. This
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -968,8 +1111,7 @@ class LogFormatter(Formatter):
                  minor_thresholds=None,
-                 linthresh=None):
-
+                 linthresh=None, linscale=None):
```
Should (the new one at least) be keyword-only?

schtandard replied:
To me it would be rather confusing to have one of them be a keyword-only parameter but not the other. If you prefer it, I'll defer to your judgement, of course.

Maybe it is worth considering to make *all* of the `Locator` and `Formatter` parameters keyword-only (going through a deprecation cycle) in a follow-up PR? This could be combined with the [suggested removal](https://github.com/matplotlib/matplotlib/pull/27310#discussion_r3509025844) of the `transform` parameter in `SymmetricalLogLoator` as well as cleaning up some inconsistencies between the signatures of different classes and methods.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -2650,158 +2810,249 @@ def nonsingular(self, vmin, vmax):
-            The multiples of integer powers of the base where ticks are placed,
-            i.e., ticks are placed at
-            ``[sub * base**i for i in ... for sub in subs]``.
+            If set, defines *base*, *lintresh* and *linscale* of the symlog transform.
```
```suggestion
            If set, defines *base*, *linthresh* and *linscale* of the symlog transform.
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -2650,158 +2810,249 @@ def nonsingular(self, vmin, vmax):
+            The maximum number of ticks to allow on a given axis. The default of
+            ``None`` will try to choose intelligently as long as this Locator has
+            already been assigned to an axis using `~.axis.Axis.get_tick_space`, but
+            otherwise falls back to 9.
```
```suggestion
            ``None`` will try to choose intelligently using
            `~.axis.Axis.get_tick_space`, as long as this Locator has already been
            assigned to an axis but otherwise falls back to 9.
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1032,43 +1173,48 @@ def set_locs(self, locs=None):
+            if self._firstsublabels == {0} and -first < vmin < vmax < first:
+                # No minor ticks are being labeled right now and the only major tick is
+                # at 0. This means the axis scaling cannot be read from the labels.
+                numsteps = int(np.ceil(first / max(-vmin, vmax)))
```
While the symmetric log scale is _intended_ for ranges that span 0, is it _guaranteed_ here that `vmin<0` and `vmax>0`, as seems to be assumed by this `max(-vmin, vmax)` call, or should this use `abs`?

schtandard replied:
No, it is not guaranteed that `vmin < 0` and `vmax > 0`, but this expression still works, as `vmin < vmax` *is* guaranteed: `vmin` can only have the greater absolute value if it is negative, `vmax` only if it is positive.

We could still convert to `max(abs(vmin), abs(vmax))` if you prefer.

QuLogic replied:
OK, I've thought about this a bit more and with `vmin < vmax`, I see what you are saying. There shouldn't be any need to change it.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -2650,158 +2810,249 @@ def nonsingular(self, vmin, vmax):
+                # Calculate the largest possible stride
+                # resulting in n_request ticks while ticking 0.
+                stride = math.ceil(n_avail / (n_request - 1)) - 1
+                if stride < n_avail / n_request:
+                    stride = n_avail // n_request
```
Is it important that this uses the two different forms of division? Otherwise, it could be one:
```suggestion
                stride = max(stride, n_avail // n_request)
```

schtandard replied:
I believe that would work. This is identical to [the calculations in `LogLocator`](https://github.com/matplotlib/matplotlib/pull/27310/changes/BASE..ec3228008b728ce1db68314a95b3e278b34d511d#diff-ffc88dbb928475229d0fc50ae389ef3f0dabc9c5b8abf42b6eb7a0af28afdb66L2576), where this test reflects the inequality in the preceding comment. If we want to change it, we should change it in both places, but I think reducing it to the single `max` would obscure the connection to the commented calculations somewhat.

schtandard committed:
Fix rst markup and typos

Co-authored-by: Elliott Sales de Andrade <quantum.analyst@gmail.com>

schtandard committed:
Improve docstring phrasing

schtandard committed:
Document all possible values of subs

schtandard committed:
Use mpl20 style in modified test

schtandard committed:
Optimize modified png images using oxipng

schtandard committed:
Clarify release note

schtandard committed:
Add example plot

schtandard commented:
@QuLogic What's the status here? Do you have further comments on the open points?

QuLogic commented:
Sorry I lost track of this; let me see what's left to do here.

schtandard commented:
While preparing [my last plot](https://github.com/matplotlib/matplotlib/pull/27310#discussion_r3888620948), I stumbled hard over #26280 and I'd like to fix it in the same release as this PR. Some coordination may be in order:

- That issue affects the same tests as this one, so maybe we should just add the fix to #26280 here? This will be less clear in principle (two distinct issues fixed in one PR), but it will avoid additional test output images in the commit graph.
- While I am hesitant to change the code here so close to the finish line: If we go with `expected_transform_b()` in #26280, that would make `SymmetricalLogTransform.transform_non_affine()` equal to `_SymmetricalLogMixin._val2axpos()`, so there might be an argument for dropping the latter and just using the former. We could also keep the duplication, arguing that it makes the code clearer and avoids depending on an implementation detail of `transform_non_affine()`, though.

What do you think?
